### PR TITLE
4.x: Improve Collect() internals

### PR DESCRIPTION
**Pull request by akarnokd**
This PR improves the internals of the `Collect()` operator (also suggested by a comment in it). See the various justifications inline.

### CODE REVIEW

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs [hunk 1]
```diff
@@ -18,50 +18,39 @@ public Collect(IObservable<TSource> source, Func<TResult> getInitialCollector, F
             _getNewCollector = getNewCollector;
         }
 
-        protected override PushToPullSink<TSource, TResult> Run(IDisposable subscription)
-        {
-            var sink = new _(this, subscription);
-            sink.Initialize();
-            return sink;
-        }
+        protected override PushToPullSink<TSource, TResult> Run(IDisposable subscription) => new _(_merge, _getNewCollector, _getInitialCollector(), subscription);
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs]: As the comment suggested, disconnect the parent from the sink and simply pass on the parameter functions as well as generate the initial value via direct call.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs [hunk 2]
```diff
@@ -18,50 +18,39 @@ public Collect(IObservable<TSource> source, Func<TResult> getInitialCollector, F
...
                 : base(subscription)
             {
-                _parent = parent;
+                _gate = new object();
+                _merge = merge;
+                _getNewCollector = getNewCollector;
+                _collector = collector;
             }
 
-            private object _gate;
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs]: `_gate` should have been readonly, moved to top.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs [hunk 3]
```diff
@@ -18,50 +18,39 @@ public Collect(IObservable<TSource> source, Func<TResult> getInitialCollector, F
...
             private TResult _collector;
-            private bool _hasFailed;
             private Exception _error;
             private bool _hasCompleted;
             private bool _done;
 
-            public void Initialize()
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs]: No need to have an initialization method when only the result of the `_getInitialCollector()` is actually needed by the sink.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs [hunk 4]
```diff
@@ -18,50 +18,39 @@ public Collect(IObservable<TSource> source, Func<TResult> getInitialCollector, F
...
-                _parent = parent;
+                _gate = new object();
+                _merge = merge;
+                _getNewCollector = getNewCollector;
+                _collector = collector;
             }
 
-            private object _gate;
             private TResult _collector;
-            private bool _hasFailed;
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs]: Failure is when `_error` is not null, no need to have a separate bool field for it as well.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs [hunk 5]
```diff
@@ -93,18 +81,19 @@ public override bool TryMoveNext(out TResult current)
             {
                 lock (_gate)
                 {
-                    if (_hasFailed)
+                    var error = _error;
+                    if (error != null)
                     {
-                        current = default(TResult);
-                        _error.Throw();
+                        current = default;
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs]: The `_collector` should probably be cleared here too, but the original code didn't do it for some reason. Should I add it?

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs [hunk 6]
```diff
@@ -93,18 +81,19 @@ public override bool TryMoveNext(out TResult current)
...
+                        error.Throw();
                     }
                     else
                     {
                         if (_hasCompleted)
                         {
                             if (_done)
                             {
-                                current = default(TResult);
+                                current = default;
```
Comment (akarnokd) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs]: The `_collector` should probably be cleared here too, but the original code didn't do it for some reason. Should I add it?

**Comment by akarnokd** [comment]
Note that a newer PR, #676, also changes this file so that PR should take priority if the changes are accepted.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs (hunk 5 above)
Comment (danielcweber) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs]: Yes

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs (hunk 6 above)
Comment (danielcweber) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs]: Yes

Comment (akarnokd) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/Collect.cs]: I'll close this and change #676.

**Pull request closed without merging**